Attempt 2 which addresses presented concerns in pr#64

dancollins34 opened the pull request:
This is a second attempt to reorganize this project so that it both can lazily load responses for interacting with the database using asynchronous requests and maintain the same api reference.

Something I've realized is that there are a lot of cases where the simplest way to organize the project for development diverges with the existing api, which is intended to be concise for use. So I've tried splitting up the two through creation of the arango_internals package. This passes all the original tests without changing imports at all. The changes which existed in the last pr to user management and async jobs still exist.

All of the changes suggested in the original pr, with the exception of the relative imports in __init__.py files which I believe are not an issue, have been implemented.

Thoughts?

joowani commented:
Hi @dancollins34,

Thank you for the PR. I am more than happy to accept all changes around lazy loading for futures, but I am still confused on why it is necessary to change the project structure. Could you provide me specific examples where the current structure hinders development? 

Also, not breaking tests isn't enough to guarantee backwards compatibility. I was only mentioning tests as a rule-of-thumb sort of thing. We have to make sure ALL imports in users' code still work, and the tests cover only a small set of possible import paths. 

I'm sorry if I wasn't clear the first time, but I am *very* hesitant on making any huge structural changes to the project. The arango_internal (arguably) seems to make things even more confusing for new devs to contribute as it now makes it more difficult to intuitively connect import paths to the file structure. Breaking up the exceptions into different modules and editing the __init__.py file to keep the import paths intact, however, is definitely a good idea and I have no problems with that.

I am not 100% against breaking backwards compatibility. I am willing to break it if there is enough justification. If you could give me some concrete scenarios it would help a lot. I am currently out of the country so I will try to provide more feedback after I return on Jan 10th. Thank you!

Best,
Joohwan

dancollins34 commented:
Hi @joowani,

Okay, I understand completely.  Here was my reasoning for doing the reorganization in the way that I did.  In the original structure, everything is laid out in a single folder.  While this does make the code which imports it shorter, imports are not typically a large part of the codebase or work.  In addition, there was a lot of duplication of functionality in the current structure; the Client object for example is a direct superset of the Database object, but does not extend it, instead has a copy of every function.  There were many cases of duplicated code, and the first thing I did was to create a unified sub-object in the same manner as was done with collections and http-clients which all the other objects could extend from.

At this point however, it became odd to work with because the related objects, BaseJob, AsyncJob and BatchJob for example, were not in the same folder and, in addition, two of them were secondary classes in files with different names.  So I tried to group the modified classes, which now inherited from the same subclasses, in folders corresponding to what they did.  However, as you pointed out in the first pr, that would have been a restructuring of any import statements and immediately cause any code manipulating the internals (i.e. not using the client object) to break.

Perhaps something I could have done instead was to just have many "from x import *" lines in the top level init.py file.  That would prevent the need for the secondary internals structure, and maintain the backwards compatibility.

-Dan

joowani commented:
Hi @dancollins34,

Right. The duplication of code in `ArangoClient` and `Database` was due to "historical reasons". The code should've been in `Database` class only, but it was initially misplaced in `ArangoClient`, and I kept it in there to maintain backward compatibility. I was planning to deprecate the code in `ArangoClient` one day, which is why you see the lazy copy-paste instead of inheritance. 

As for `BatchJob` and `AsyncJob`, I think you can use your `BaseJob` and still keep backward compatibility by importing the Job classes from new locations in batch.py and async.py.

I do agree that it is worth re-evaluating the whole project structure, but again I want to avoid breaking user code at all cost. I think this is something we need to approach more carefully.

At this point, I believe it would be best to separate the pull requests into two. First one would *only* contain your futures changes. I would like to get these unblocked and merged asap. Once that's in, I would like to do a comprehensive review of the project and make changes where necessary (and also add support for the new version of ArangoDB). After that, we can revisit the structure overhaul. How does that sound? 

Best,
Joohwan

dancollins34 commented:
Hi @joowani,

So this is my final attempt on this current structure, only because a full redo of what I've done would be a significant amount of time and effort.  If you don't like the current version then I'll start on what you suggested.

I've removed the internals directory and added the required renames to account for the file moves.  And what I've realized was that, after removing the API directory, there are only 2 __init__ references which need to be added to maintain compatibility with the old api- arango.aql.AQL and arango.graph.Graph , everything else works because it is now in a package with the same name as its original file.  If you would still prefer I start from scratch on this, I can, but I'd ask you to take a look at this first.

-Dan

joowani commented on arango/connections/base.py:
```diff
@@ -0,0 +1,259 @@
+        return self._async_ready
+
+    def handle_request(self, request, handler, job_class=None,
+                       **kwargs):
```
Please use explicit arguments instead of `**kwargs`.

dancollins34 replied:
The kwargs are used exclusively for passing arguments for the job constructor. I could create an argument struct instead for this purpose, would that work?

joowani replied:
Which arguments are they exactly?

dancollins34 replied:
response, connection, and return_result for AsyncJob, response, job_id, assign_id, and job_type for BaseJob, job_id and response for BatchJob, and job_id and response for SynchronousResultJob

joowani commented on arango/cursors/base.py:
```diff
@@ -5,25 +5,30 @@
 
-    def __init__(self, connection, init_data):
-        self._conn = connection
+    def __init__(self, connection, init_data, cursor_type="cursor"):
```
I see that for jobs we pass in the class, but for cursor types we pass in magic string. This seems inconsistent. Also, why do we have to pass in such information using a new argument? Have you considered using `isinstance`?

dancollins34 replied:
The string itself used in the format string is the only api difference between the two cursors, they otherwise behave identically on the client side. We could create a wrapper class, but that adds a lot of extra complexity for a single string change

joowani replied:
Doesn't this also break backwards compatibility? `Cursor` and `ExportCursor` should remain available to the users.

dancollins34 replied:
Okay, my first comment was incorrect, I forgot I had kept the ExportCursor class.  I'll use a setter instead of having this in the init.

joowani commented on arango/cursors/base.py:
```diff
@@ -118,55 +123,28 @@ def next(self):
-class ExportCursor(Cursor):  # pragma: no cover
-    """ArangoDB cursor for export queries only.
+            job = self.handle_request(request, handler, job_class=BaseJob,
+                                      use_underlying=True)
```
I still do not like the argument name `user_underlying` as its too vague and I cannot figure out what it does.

dancollins34 replied:
use_underlying indicates whether the request should use the connection that created it or the underlying connection for that connection.  Since the underlying property recursively checks the underlying of its parents, this allows for things like having an asycnhronous execution wrapping a batch execution, while still allowing the AsyncJobs to use the underlying BaseConnection object to get their status, instead of using the batch connection

dancollins34 replied:
If you have a better name for this behavior, let me know

joowani commented on tests/test_batch.py:
```diff
@@ -139,7 +137,10 @@ def test_batch_insert_context_manager_no_commit_on_error():
         assert len(col) == 0
         assert job1.status() == 'pending'
-        assert job1.result() is None
+        # TODO CHANGED Behavior: Result of jobs without a response fails on
```
There seems to be a lot of comments hanging around (e.g. TODO CHANGED). Please review them again clean up if possible.

dancollins34 replied:
The “TODO CHANGED” markers were to indicate the slight behavioral changes (mostly throwing a different exception) I made in the tests. If you are okay with all the changes where they are marked, I will remove the comments.

joowani replied:
Actually.... I think it would be better to leave the comments for now so I can track the changes.

dancollins34 replied:
Yep. Let me know when you want me to remove them

dancollins34 commented on arango/__init__.py:
```diff
@@ -1,2 +1,4 @@
-from arango.client import ArangoClient  # noqa: F401
-from arango.exceptions import ArangoError  # noqa: F401
+from .api import APIWrapper
```
Sure. I’ll make all imports absolute if necessary

dancollins34 commented on arango/collections/base.py:
```diff
@@ -167,7 +208,7 @@ def rename(self, new_name):
         request = Request(
             method='put',
-            endpoint='/_api/collection/{}/rename'.format(self._name),
+            url='/_api/collection/{}/rename'.format(self._name),
```
Will undo all these renames

dancollins34 commented on arango/collections/base.py:
```diff
@@ -689,18 +736,24 @@ def find_near(self, latitude, longitude, limit=None):
         """
+
+        if limit is None:
+            limit_string = ""
```
I tend to use double quotes because they are consistent Unicode python 2->3, whereas single quotes are byte strings in python 2 and Unicode in 3, but I can switch to single if you’d like

joowani replied:
Oh hmm I didn't know that (good to know). You can leave them as they are then.

dancollins34 replied:
Actually, perhaps I'm wrong on this one? Everywhere seems to indicate that they're exactly the same, but I remember having handling issues when working with lxml...

Anyway, for what we're doing, the difference shouldn't matter all that much since all of our constants are ascii.  I'll use single quotes.

dancollins34 commented on tests/test_version.py:
```diff
@@ -1,6 +1,6 @@
-from arango.version import VERSION
+from arango.version import __VERSION__
```
This was named as it was because of a comment you made in the previous pr. Which would you prefer?

joowani replied:
Did I say `__VERSION__` before? If I did sorry about that. Please change it to lowercase as that seems to be the general practice.

dancollins34 replied:
will do

dancollins34 commented:
No idea why I can't respond to the main comment. Anyway:

equests:

Please leave the permissions of the files as they are (644). I'm not sure if it was intentional, but I see some have changed.

I didn't know git persisted permissions.  I'll run a chmod on everything before pushing again.

Please undo misc. refactoring of variable/argument names as they can really bloat up the PR (e.g. endpoint -> url).

Yep

I know it's not consistent, but for backwards compatibility, please undo any non-critical renames for now (e.g. Cursor -> BaseCursor, Connection -> BaseConnection, Transaction -> TransactionExecution).

Yep

Use absolute paths instead of relative ones.

Yep

Change **kwargs to explicit arguments.

See specific comment.  Could we use a config struct?

One of the goals of python-arango is to abstract away the ArangoDB's rather awkward concept of "system database". Introducing SystemDatabase seems like a step backwards in this sense.

I could move the "SystemDatabase" functionality into the "Client" functionality, and then just have Client extend Database

I'm not sure if the arango/connections/executions directory is necessary. Why can't we just place them under arango/connections?

Sounds good.

joowani commented:
Yea github has some limitations.... :(

re: I could move the "SystemDatabase" functionality into the "Client" functionality, and then just have Client extend Database
Isn't that essentially how it was originally (minus the inheritance)? For now I would like to leave ArangoClient and Database as they are.
Thanks!


________________________________
From: dancollins34 <notifications@github.com>
Sent: January 15, 2018 2:15:58 PM
To: joowani/python-arango
Cc: Joohwan; Mention
Subject: Re: [joowani/python-arango] Attempt 2 which addresses presented concerns in pr#64 (#66)


No idea why I can't respond to the main comment. Anyway:

equests:

Please leave the permissions of the files as they are (644). I'm not sure if it was intentional, but I see some have changed.

I didn't know git persisted permissions. I'll run a chmod on everything before pushing again.

Please undo misc. refactoring of variable/argument names as they can really bloat up the PR (e.g. endpoint -> url).

Yep

I know it's not consistent, but for backwards compatibility, please undo any non-critical renames for now (e.g. Cursor -> BaseCursor, Connection -> BaseConnection, Transaction -> TransactionExecution).

Yep

Use absolute paths instead of relative ones.

Yep

Change **kwargs to explicit arguments.

See specific comment. Could we use a config struct?

One of the goals of python-arango is to abstract away the ArangoDB's rather awkward concept of "system database". Introducing SystemDatabase seems like a step backwards in this sense.

I could move the "SystemDatabase" functionality into the "Client" functionality, and then just have Client extend Database

I'm not sure if the arango/connections/executions directory is necessary. Why can't we just place them under arango/connections?

Sounds good.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/joowani/python-arango/pull/66#issuecomment-357801277>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACk6cv8uwH0NxFAsc_Dxeyk6TYbpuWsoks5tK84egaJpZM4RPda4>.

dancollins34 commented:
yes, that was how it was originally, minus inheritance.  I'll do that.  Although, the restriction on that would be you wouldn't be able to call client.db("_system").endpoints() for example.

joowani commented:
Hi @dancollins34, 

Overall this PR's looking pretty good. It's come a long way (and it's been dragging on for a while) so I will merge it as is. I still see some issues here and there but they seem like quick fixes so I will apply them myself. Please note that I will rebasing and overwriting the history once again so you will need to checkout a fresh copy of dev at some point. Thanks again for your contribution!

dancollins34 commented:
Okay, I was working on fixing the kwargs and double quotes -> single last night, there’s just a lot to do. I’ll submit that in a separate pr

joowani commented:
Sure. Thanks @dancollins34.
